File: rust/src/streaming/service.rs
// SPDX-License-Identifier: MIT

//! Live streaming service: announce streams, subscribe, publish segments.

use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

use crate::constants::DEFAULT_TTL;
use crate::protocol::{MeshPacket, PacketType};
use crate::routing::sender::MeshSender;
use crate::routing::RoutingService;

// ── Wire types ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StreamAnnouncePayload {
    pub stream_id: Uuid,
    pub title: String,
    pub content_type: String,
    pub codec: String,
    pub segment_duration_ms: u64,
    pub state: String,
    pub started_at_ms: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StreamSubscribePayload {
    pub stream_id: Uuid,
    pub live_only: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StreamUnsubscribePayload {
    pub stream_id: Uuid,
}

// ── Stream state ───────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct StreamEntry {
    pub stream_id: Uuid,
    pub title: String,
    pub content_type: String,
    pub codec: String,
    pub segment_duration_ms: u64,
    pub started_at_ms: i64,
    pub active: bool,
}

// ── Service ───────────────────────────────────────────────────────────────────

pub struct StreamingService {
    sender: Arc<dyn MeshSender>,
    routing: Arc<RoutingService>,
    /// Streams this node is publishing.
    streams: Mutex<HashMap<Uuid, StreamEntry>>,
    /// stream_id -> set of subscriber UHIDs.
    subscribers: Mutex<HashMap<Uuid, HashSet<String>>>,
}

impl StreamingService {
    pub fn new(sender: Arc<dyn MeshSender>, routing: Arc<RoutingService>) -> Self {
        Self {
            sender,
            routing,
            streams: Mutex::new(HashMap::new()),
            subscribers: Mutex::new(HashMap::new()),
        }
    }

    /// Start a new stream. Returns the stream_id.
    pub async fn start_stream(
        &self,
        title: &str,
        content_type: &str,
        codec: &str,
        segment_duration_ms: u64,
    ) -> Result<Uuid, String> {
        let stream_id = Uuid::new_v4();
        let now = unix_millis();
        let entry = StreamEntry {
            stream_id,
            title: title.to_string(),
            content_type: content_type.to_string(),
            codec: codec.to_string(),
            segment_duration_ms,
            started_at_ms: now,
            active: true,
        };
        {
            let mut streams = self.streams.lock().unwrap();
            streams.insert(stream_id, entry.clone());
        }
        let ann = StreamAnnouncePayload {
            stream_id,
            title: entry.title.clone(),
            content_type: entry.content_type.clone(),
            codec: entry.codec.clone(),
            segment_duration_ms,
            state: "live".into(),
            started_at_ms: now,
        };
        self.broadcast_announce(&ann).await;
        Ok(stream_id)
    }

    /// End a stream and notify subscribers.
    pub async fn end_stream(&self, stream_id: Uuid) -> Result<(), String> {
        let entry = {
            let mut streams = self.streams.lock().unwrap();
            let entry = streams
                .get_mut(&stream_id)
                .ok_or_else(|| format!("unknown stream_id {}", stream_id))?;
            entry.active = false;
            entry.clone()
        };
        let ann = StreamAnnouncePayload {
            stream_id,
            title: entry.title,
            content_type: entry.content_type,
            codec: entry.codec,
            segment_duration_ms: entry.segment_duration_ms,
            state: "ended".into(),
            started_at_ms: entry.started_at_ms,
        };
        self.broadcast_announce(&ann).await;
        Ok(())
    }

    /// Subscribe to a stream published by `publisher_uhid`.
    pub async fn subscribe(
        &self,
        stream_id: Uuid,
        publisher_uhid: &str,
        live_only: bool,
    ) -> Result<(), String> {
        if publisher_uhid.is_empty() {
            return Err("publisher_uhid is empty".into());
        }
        {
            let mut subs = self.subscribers.lock().unwrap();
            subs.entry(stream_id)
                .or_insert_with(HashSet::new)
                .insert(self.sender.local_uhid());
        }
        let body = serde_json::to_vec(&StreamSubscribePayload { stream_id, live_only })
            .unwrap_or_default();
        let route = self.routing.find_route(publisher_uhid).await;
        let next_hop = route
            .map(|r| r.next_hop_uhid)
            .unwrap_or_else(|| publisher_uhid.to_string());

        let mut pkt = MeshPacket::new(PacketType::StreamSubscribe, self.sender.local_uhid());
        pkt.destination_uhid = publisher_uhid.to_string();
        pkt.ttl = DEFAULT_TTL;
        pkt.payload = body;
        self.sender.send(&pkt, &next_hop).await;
        Ok(())
    }

    /// Unsubscribe from a stream.
    pub async fn unsubscribe(
        &self,
        stream_id: Uuid,
        publisher_uhid: &str,
    ) -> Result<(), String> {
        {
            let mut subs = self.subscribers.lock().unwrap();
            if let Some(set) = subs.get_mut(&stream_id) {
                set.remove(&self.sender.local_uhid());
            }
        }
        let body = serde_json::to_vec(&StreamUnsubscribePayload { stream_id })
            .unwrap_or_default();
        let route = self.routing.find_route(publisher_uhid).await;
        let next_hop = route
            .map(|r| r.next_hop_uhid)
            .unwrap_or_else(|| publisher_uhid.to_string());

        let mut pkt = MeshPacket::new(PacketType::StreamUnsubscribe, self.sender.local_uhid());
        pkt.destination_uhid = publisher_uhid.to_string();
        pkt.ttl = DEFAULT_TTL;
        pkt.payload = body;
        self.sender.send(&pkt, &next_hop).await;
        Ok(())
    }

    /// Publish a segment to all current subscribers.
    ///
    /// Binary payload wire format:
    ///   [16] StreamId (UUID RFC4122 big-endian)
    ///   [4]  Sequence (u32 little-endian)
    ///   [8]  TimestampMs (i64 little-endian)
    ///   [1]  IsKeyframe (0 or 1)
    ///   [N]  EncodedPayload
    pub async fn publish_segment(
        &self,
        stream_id: Uuid,
        data: &[u8],
        is_keyframe: bool,
    ) -> Result<(), String> {
        {
            let streams = self.streams.lock().unwrap();
            if !streams
                .get(&stream_id)
                .map(|s| s.active)
                .unwrap_or(false)
            {
                return Err(format!("stream {} is not active", stream_id));
            }
        }

        let subs: Vec<String> = {
            let subs = self.subscribers.lock().unwrap();
            subs.get(&stream_id)
                .cloned()
                .unwrap_or_default()
                .into_iter()
                .collect()
        };

        let payload = build_stream_segment(stream_id, data, is_keyframe);
        let local = self.sender.local_uhid();

        for uhid in &subs {
            let mut pkt = MeshPacket::new(PacketType::StreamSegment, local.clone());
            pkt.destination_uhid = uhid.clone();
            pkt.ttl = DEFAULT_TTL;
            pkt.priority = 64;
            pkt.payload = payload.clone();
            self.sender.send(&pkt, uhid).await;
        }
        Ok(())
    }

    /// Route an inbound packet to the right handler.
    pub async fn handle_packet(&self, packet: &MeshPacket) -> Result<(), String> {
        match packet.packet_type {
            PacketType::StreamAnnounce => self.handle_announce(packet).await,
            PacketType::StreamSubscribe => self.handle_subscribe(packet).await,
            PacketType::StreamUnsubscribe => self.handle_unsubscribe(packet).await,
            PacketType::StreamSegment => Ok(()), // surfaced to app layer
            _ => Ok(()),
        }
    }

    // ── private ────────────────────────────────────────────────────────────────

    async fn broadcast_announce(&self, ann: &StreamAnnouncePayload) {
        let body = match serde_json::to_vec(ann) {
            Ok(b) => b,
            Err(_) => return,
        };
        let mut pkt = MeshPacket::new(PacketType::StreamAnnounce, self.sender.local_uhid());
        pkt.destination_uhid = String::new();
        pkt.ttl = DEFAULT_TTL;
        pkt.payload = body;
        self.sender.broadcast(&pkt).await;
    }

    async fn handle_announce(&self, _packet: &MeshPacket) -> Result<(), String> {
        // Announcements from remote publishers are surfaced to the app layer.
        Ok(())
    }

    async fn handle_subscribe(&self, packet: &MeshPacket) -> Result<(), String> {
        let sub: StreamSubscribePayload = serde_json::from_slice(&packet.payload)
            .map_err(|e| format!("bad subscribe payload: {}", e))?;
        let mut subs = self.subscribers.lock().unwrap();
        subs.entry(sub.stream_id)
            .or_insert_with(HashSet::new)
            .insert(packet.source_uhid.clone());
        Ok(())
    }

    async fn handle_unsubscribe(&self, packet: &MeshPacket) -> Result<(), String> {
        let unsub: StreamUnsubscribePayload = serde_json::from_slice(&packet.payload)
            .map_err(|e| format!("bad unsubscribe payload: {}", e))?;
        let mut subs = self.subscribers.lock().unwrap();
        if let Some(set) = subs.get_mut(&unsub.stream_id) {
            set.remove(&packet.source_uhid);
        }
        Ok(())
    }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

static SEG_SEQ: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);

fn build_stream_segment(stream_id: Uuid, data: &[u8], is_keyframe: bool) -> Vec<u8> {
    let seq = SEG_SEQ.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    let ts = unix_millis();
    let mut buf = Vec::with_capacity(29 + data.len());
    buf.extend_from_slice(stream_id.as_bytes()); // [16]
    buf.extend_from_slice(&seq.to_le_bytes()); // [4]
    buf.extend_from_slice(&ts.to_le_bytes()); // [8]
    buf.push(if is_keyframe { 1 } else { 0 }); // [1]
    buf.extend_from_slice(data); // [N]
    buf
}

fn unix_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

File: go/streaming/service.go
// SPDX-License-Identifier: MIT

// Package streaming implements live stream publishing, subscription, and segment
// relay for the Aether mesh. VideoCallService and WatchTogetherService are in
// the same package.
//
// Wire formats:
//
//	StreamAnnounce    → JSON (StreamAnnouncePayload)
//	StreamSubscribe   → JSON (StreamSubscribePayload)
//	StreamUnsubscribe → JSON (StreamUnsubscribePayload)
//	StreamSegment     → binary StreamSegment
package streaming

import (
	"bytes"
	"context"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/bhengubv/aether-protocol/go/constants"
	"github.com/bhengubv/aether-protocol/go/protocol"
	"github.com/bhengubv/aether-protocol/go/routing"
)

// ──────────────────────────────────────────────
// JSON wire types
// ──────────────────────────────────────────────

// StreamAnnouncePayload is the JSON body for StreamAnnounce packets.
type StreamAnnouncePayload struct {
	StreamID          string `json:"stream_id"`
	Title             string `json:"title"`
	ContentType       string `json:"content_type"`
	Codec             string `json:"codec"`
	SegmentDurationMs int    `json:"segment_duration_ms"`
	State             string `json:"state"` // "live" | "ended"
	StartedAtMs       int64  `json:"started_at_ms"`
}

// StreamSubscribePayload is the JSON body for StreamSubscribe packets.
type StreamSubscribePayload struct {
	StreamID string `json:"stream_id"`
	LiveOnly bool   `json:"live_only"`
}

// StreamUnsubscribePayload is the JSON body for StreamUnsubscribe packets.
type StreamUnsubscribePayload struct {
	StreamID string `json:"stream_id"`
}

// ──────────────────────────────────────────────
// Local session state
// ──────────────────────────────────────────────

// StreamSession holds runtime state for an active outbound stream.
type StreamSession struct {
	StreamID          uuid.UUID
	Title             string
	ContentType       string
	Codec             string
	SegmentDurationMs int
	Sequence          uint32
	StartedAt         time.Time
	subscribers       sync.Map // uhid -> struct{}
}

// ──────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────

// StreamingService manages live audio/video stream publishing and relaying.
type StreamingService struct {
	sender    routing.MeshSender
	localUhid string
	streams   sync.Map // uuid.UUID -> *StreamSession

	// subscribers maps streamID -> *sync.Map (uhid -> struct{})
	// used by the relay side to track remote subscribers.
	subscribers sync.Map

	// Event callbacks — optional.
	OnStreamAnnounced    func(payload *StreamAnnouncePayload, fromUhid string)
	OnStreamEnded        func(streamID uuid.UUID, fromUhid string)
	OnSegmentReceived    func(streamID uuid.UUID, segment *StreamSegmentFrame)
	OnSubscribed         func(streamID uuid.UUID, fromUhid string)
	OnUnsubscribed       func(streamID uuid.UUID, fromUhid string)
}

// StreamSegmentFrame is the decoded form of a binary StreamSegment payload.
type StreamSegmentFrame struct {
	StreamID       uuid.UUID
	Sequence       uint32
	TimestampMs    int64
	IsKeyframe     bool
	EncodedPayload []byte
}

// NewStreamingService constructs a StreamingService.
func NewStreamingService(sender routing.MeshSender) *StreamingService {
	if sender == nil {
		panic("streaming: sender must not be nil")
	}
	return &StreamingService{
		sender:    sender,
		localUhid: sender.LocalUhid(),
	}
}

// StartStream announces a new live stream on the mesh. Returns the new stream ID.
func (s *StreamingService) StartStream(ctx context.Context, title, contentType, codec string, segmentDurationMs int) (uuid.UUID, error) {
	if title == "" {
		return uuid.Nil, errors.New("streaming: title must not be empty")
	}
	if segmentDurationMs <= 0 {
		segmentDurationMs = int(constants.DefaultSegmentDurationMs)
	}

	streamID := uuid.New()
	session := &StreamSession{
		StreamID:          streamID,
		Title:             title,
		ContentType:       contentType,
		Codec:             codec,
		SegmentDurationMs: segmentDurationMs,
		StartedAt:         time.Now(),
	}
	s.streams.Store(streamID, session)

	payload := StreamAnnouncePayload{
		StreamID:          streamID.String(),
		Title:             title,
		ContentType:       contentType,
		Codec:             codec,
		SegmentDurationMs: segmentDurationMs,
		State:             "live",
		StartedAtMs:       session.StartedAt.UnixMilli(),
	}
	return streamID, s.broadcastAnnounce(ctx, payload)
}

// EndStream stops a stream and notifies subscribers.
func (s *StreamingService) EndStream(ctx context.Context, streamID uuid.UUID) error {
	session, err := s.getStream(streamID)
	if err != nil {
		return err
	}
	s.streams.Delete(streamID)

	payload := StreamAnnouncePayload{
		StreamID:    streamID.String(),
		State:       "ended",
		StartedAtMs: session.StartedAt.UnixMilli(),
	}
	return s.broadcastAnnounce(ctx, payload)
}

// Subscribe sends a StreamSubscribe packet to publisherUhid.
func (s *StreamingService) Subscribe(ctx context.Context, streamID uuid.UUID, publisherUhid string, liveOnly bool) error {
	if publisherUhid == "" {
		return errors.New("streaming: publisherUhid must not be empty")
	}
	body, err := json.Marshal(StreamSubscribePayload{
		StreamID: streamID.String(),
		LiveOnly: liveOnly,
	})
	if err != nil {
		return fmt.Errorf("streaming: marshal subscribe: %w", err)
	}
	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.StreamSubscribe
	pkt.SourceUhid = s.localUhid
	pkt.DestinationUhid = publisherUhid
	pkt.Ttl = constants.DefaultTtl
	pkt.Payload = body
	_, err = s.sender.Send(ctx, pkt, publisherUhid)
	return err
}

// Unsubscribe sends a StreamUnsubscribe packet to publisherUhid.
func (s *StreamingService) Unsubscribe(ctx context.Context, streamID uuid.UUID, publisherUhid string) error {
	if publisherUhid == "" {
		return errors.New("streaming: publisherUhid must not be empty")
	}
	body, err := json.Marshal(StreamUnsubscribePayload{StreamID: streamID.String()})
	if err != nil {
		return fmt.Errorf("streaming: marshal unsubscribe: %w", err)
	}
	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.StreamUnsubscribe
	pkt.SourceUhid = s.localUhid
	pkt.DestinationUhid = publisherUhid
	pkt.Ttl = constants.DefaultTtl
	pkt.Payload = body
	_, err = s.sender.Send(ctx, pkt, publisherUhid)
	return err
}

// PublishSegment encodes and delivers a segment to all current subscribers.
//
// Binary layout (StreamSegment):
//
//	[16] StreamId      (UUID, big-endian)
//	[4]  Sequence      (uint32, little-endian)
//	[8]  TimestampMs   (int64, little-endian)
//	[1]  IsKeyframe    (0 or 1)
//	[N]  EncodedPayload
func (s *StreamingService) PublishSegment(ctx context.Context, streamID uuid.UUID, data []byte, isKeyframe bool) error {
	session, err := s.getStream(streamID)
	if err != nil {
		return err
	}
	session.Sequence++
	payload, err := marshalStreamSegment(streamID, session.Sequence, time.Now().UnixMilli(), isKeyframe, data)
	if err != nil {
		return err
	}

	// Fan out to all registered subscribers for this stream.
	subMap := s.getOrCreateSubMap(streamID)
	subMap.Range(func(key, _ any) bool {
		peerUhid, _ := key.(string)
		pkt := protocol.NewMeshPacket()
		pkt.Type = protocol.StreamSegment
		pkt.SourceUhid = s.localUhid
		pkt.DestinationUhid = peerUhid
		pkt.Ttl = constants.DefaultTtl
		pkt.Payload = payload
		_, _ = s.sender.Send(ctx, pkt, peerUhid)
		return true
	})
	return nil
}

// HandlePacket processes inbound streaming-related packets.
func (s *StreamingService) HandlePacket(ctx context.Context, packet *protocol.MeshPacket) error {
	if packet == nil {
		return errors.New("streaming: packet must not be nil")
	}
	switch packet.Type {
	case protocol.StreamAnnounce:
		return s.handleAnnounce(packet)
	case protocol.StreamSubscribe:
		return s.handleSubscribe(ctx, packet)
	case protocol.StreamUnsubscribe:
		return s.handleUnsubscribe(ctx, packet)
	case protocol.StreamSegment:
		return s.handleSegment(packet)
	default:
		return fmt.Errorf("streaming: unexpected packet type %s", packet.Type)
	}
}

// ──────────────────────────────────────────────
// Internal helpers
// ──────────────────────────────────────────────

func (s *StreamingService) handleAnnounce(packet *protocol.MeshPacket) error {
	var payload StreamAnnouncePayload
	if err := json.Unmarshal(packet.Payload, &payload); err != nil {
		return fmt.Errorf("streaming: unmarshal announce: %w", err)
	}
	switch payload.State {
	case "live":
		if cb := s.OnStreamAnnounced; cb != nil {
			cb(&payload, packet.SourceUhid)
		}
	case "ended":
		streamID, err := uuid.Parse(payload.StreamID)
		if err == nil {
			if cb := s.OnStreamEnded; cb != nil {
				cb(streamID, packet.SourceUhid)
			}
		}
	}
	return nil
}

func (s *StreamingService) handleSubscribe(_ context.Context, packet *protocol.MeshPacket) error {
	var payload StreamSubscribePayload
	if err := json.Unmarshal(packet.Payload, &payload); err != nil {
		return fmt.Errorf("streaming: unmarshal subscribe: %w", err)
	}
	streamID, err := uuid.Parse(payload.StreamID)
	if err != nil {
		return fmt.Errorf("streaming: invalid stream_id: %w", err)
	}
	subMap := s.getOrCreateSubMap(streamID)
	subMap.Store(packet.SourceUhid, struct{}{})
	if cb := s.OnSubscribed; cb != nil {
		cb(streamID, packet.SourceUhid)
	}
	return nil
}

func (s *StreamingService) handleUnsubscribe(_ context.Context, packet *protocol.MeshPacket) error {
	var payload StreamUnsubscribePayload
	if err := json.Unmarshal(packet.Payload, &payload); err != nil {
		return fmt.Errorf("streaming: unmarshal unsubscribe: %w", err)
	}
	streamID, err := uuid.Parse(payload.StreamID)
	if err != nil {
		return fmt.Errorf("streaming: invalid stream_id: %w", err)
	}
	if m, ok := s.subscribers.Load(streamID); ok {
		m.(*sync.Map).Delete(packet.SourceUhid)
	}
	if cb := s.OnUnsubscribed; cb != nil {
		cb(streamID, packet.SourceUhid)
	}
	return nil
}

func (s *StreamingService) handleSegment(packet *protocol.MeshPacket) error {
	frame, err := unmarshalStreamSegment(packet.Payload)
	if err != nil {
		return fmt.Errorf("streaming: unmarshal segment: %w", err)
	}
	if cb := s.OnSegmentReceived; cb != nil {
		cb(frame.StreamID, frame)
	}
	return nil
}

func (s *StreamingService) getStream(streamID uuid.UUID) (*StreamSession, error) {
	v, ok := s.streams.Load(streamID)
	if !ok {
		return nil, fmt.Errorf("streaming: unknown stream %s", streamID)
	}
	return v.(*StreamSession), nil
}

func (s *StreamingService) getOrCreateSubMap(streamID uuid.UUID) *sync.Map {
	v, _ := s.subscribers.LoadOrStore(streamID, &sync.Map{})
	return v.(*sync.Map)
}

func (s *StreamingService) broadcastAnnounce(ctx context.Context, payload StreamAnnouncePayload) error {
	body, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("streaming: marshal announce: %w", err)
	}
	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.StreamAnnounce
	pkt.SourceUhid = s.localUhid
	pkt.DestinationUhid = ""
	pkt.Ttl = constants.DefaultTtl
	pkt.Payload = body
	_, err = s.sender.Broadcast(ctx, pkt)
	return err
}

// ──────────────────────────────────────────────
// Binary serialisation helpers
// ──────────────────────────────────────────────

// marshalStreamSegment encodes a StreamSegment into its canonical binary format.
func marshalStreamSegment(streamID uuid.UUID, seq uint32, tsMs int64, isKeyframe bool, payload []byte) ([]byte, error) {
	var buf bytes.Buffer
	buf.Write(streamID[:])
	if err := binary.Write(&buf, binary.LittleEndian, seq); err != nil {
		return nil, err
	}
	if err := binary.Write(&buf, binary.LittleEndian, tsMs); err != nil {
		return nil, err
	}
	if isKeyframe {
		buf.WriteByte(1)
	} else {
		buf.WriteByte(0)
	}
	buf.Write(payload)
	return buf.Bytes(), nil
}

// unmarshalStreamSegment decodes a binary StreamSegment payload.
func unmarshalStreamSegment(data []byte) (*StreamSegmentFrame, error) {
	const fixedSize = 16 + 4 + 8 + 1
	if len(data) < fixedSize {
		return nil, fmt.Errorf("streaming: segment too short: %d bytes", len(data))
	}
	r := bytes.NewReader(data)

	var streamIDBytes [16]byte
	if _, err := r.Read(streamIDBytes[:]); err != nil {
		return nil, err
	}
	streamID, err := uuid.FromBytes(streamIDBytes[:])
	if err != nil {
		return nil, err
	}

	var seq uint32
	if err := binary.Read(r, binary.LittleEndian, &seq); err != nil {
		return nil, err
	}

	var tsMs int64
	if err := binary.Read(r, binary.LittleEndian, &tsMs); err != nil {
		return nil, err
	}

	kfByte, err := r.ReadByte()
	if err != nil {
		return nil, err
	}

	encoded := make([]byte, r.Len())
	if _, err := r.Read(encoded); err != nil && r.Len() > 0 {
		return nil, err
	}

	return &StreamSegmentFrame{
		StreamID:       streamID,
		Sequence:       seq,
		TimestampMs:    tsMs,
		IsKeyframe:     kfByte != 0,
		EncodedPayload: encoded,
	}, nil
}
